src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -834,42 +836,96 @@ internal unsafe int GetNonRandomizedHashCode()
         }
 
-        // Use this if and only if 'Denial of Service' attacks are not a concern (i.e. never used for free-form user input),
-        // or are otherwise mitigated
         internal unsafe int GetNonRandomizedHashCodeOrdinalIgnoreCase()
         {
+            uint hash1 = (5381 << 16) + 5381;
+            uint hash2 = hash1;
+
             fixed (char* src = &_firstChar)
             {
                 Debug.Assert(src[this.Length] == '\0', "src[this.Length] == '\\0'");
-                Debug.Assert(((int)src) % 4 == 0, "Managed string should start at 4 bytes boundary");
+                Debug.Assert(((int) src) % 4 == 0, "Managed string should start at 4 bytes boundary");
 
-                uint hash1 = (5381 << 16) + 5381;
-                uint hash2 = hash1;
-
-                uint* ptr = (uint*)src;
+                uint* ptr = (uint*) src;
                 int length = this.Length;
 
                 // We "normalize to lowercase" every char by ORing with 0x0020. This casts
                 // a very wide net because it will change, e.g., '^' to '~'. But that should
                 // be ok because we expect this to be very rare in practice.
-
                 const uint NormalizeToLowercase = 0x0020_0020u; // valid both for big-endian and for little-endian
 
                 while (length > 2)
                 {
+                    uint p0 = ptr[0];
+                    uint p1 = ptr[1];
+                    if (!Utf16Utility.AllCharsInUInt32AreAscii(p0 | p1))
+                    {
+                        goto NotAscii;
+                    }
+
                     length -= 4;
                     // Where length is 4n-1 (e.g. 3,7,11,15,19) this additionally consumes the null terminator
-                    hash1 = (BitOperations.RotateLeft(hash1, 5) + hash1) ^ (ptr[0] | NormalizeToLowercase);
-                    hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (ptr[1] | NormalizeToLowercase);
+                    hash1 = (BitOperations.RotateLeft(hash1, 5) + hash1) ^ (p0 | NormalizeToLowercase);
+                    hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (p1 | NormalizeToLowercase);
                     ptr += 2;
                 }
 
                 if (length > 0)
                 {
+                    uint p0 = ptr[0];
+                    if (!Utf16Utility.AllCharsInUInt32AreAscii(p0))
+                    {
+                        goto NotAscii;
+                    }
+
                     // Where length is 4n-3 (e.g. 1,5,9,13,17) this additionally consumes the null terminator
-                    hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (ptr[0] | NormalizeToLowercase);
+                    hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (p0 | NormalizeToLowercase);
+                }
+            }
+
+            return (int)(hash1 + (hash2 * 1566083941));
+
+        NotAscii:
+            return GetNonRandomizedHashCodeOrdinalIgnoreCaseSlow(this, hash1, hash2);
```

Comment:
I do not think this it correct to pass hash1 and hash2 into the slow method. We could have processed some number of characters already and so hash1 and hash2 may not have their original values.

It would be nice to add a test case that covers this case.

It would work only if we also passed in the index

Reply:
oops, indeed
I am not sure it can be tested anyhow other than hardcoding the expected hashcode value 🤔